--- ascat.R_2.5	2020-05-06 17:44:55.000000000 +0200
+++ ascat.R	2020-05-08 00:11:47.092920834 +0200
@@ -269,10 +269,15 @@
         }
       }
       corr<-apply(corr_tot,1,function(x) sum(abs(x*length_tot))/sum(length_tot))
-      index_1M<-c(which(names(corr)=="X1M"),which(names(corr)=="X1Mb"))
-      maxGCcol_short<-which(corr[1:(index_1M-1)]==max(corr[1:(index_1M-1)]))
-      maxGCcol_long<-which(corr[index_1M:length(corr)]==max(corr[index_1M:length(corr)]))
-      maxGCcol_long<-(maxGCcol_long+(index_1M-1))
+#      index_1M<-c(which(names(corr)=="X1M"),which(names(corr)=="X1Mb"))
+#      maxGCcol_short<-which(corr[1:(index_1M-1)]==max(corr[1:(index_1M-1)]))
+#      maxGCcol_long<-which(corr[index_1M:length(corr)]==max(corr[index_1M:length(corr)]))
+#      maxGCcol_long<-(maxGCcol_long+(index_1M-1))
+
+      index_1000<-which(names(corr)=="X1000")
+      maxGCcol_short<-which(corr[1:(index_1000-1)]==max(corr[1:(index_1000-1)]))
+      maxGCcol_long<-which(corr[index_1000:length(corr)]==max(corr[index_1000:length(corr)]))
+      maxGCcol_long<-(maxGCcol_long+(index_1000-1))
       
       cat("weighted correlation: ",paste(names(corr),format(corr,digits=2), ";"),"\n")
       cat("Short window size: ",names(GC_newlist)[maxGCcol_short+2],"\n")
